File: Planning/src/edu/uno/ai/planning/logic/Conjunction.java
package edu.uno.ai.planning.logic;

import edu.uno.ai.planning.MutableState;
import edu.uno.ai.planning.State;
import edu.uno.ai.planning.util.ImmutableArray;

/**
 * A conjunction is a Boolean expression that is true when all of its conjuncts
 * are true.
 * 
 * @author Stephen G. Ware
 */
public class Conjunction extends NAryBooleanExpression {

	/** The Boolean operator used for expressing conjunctions */
	public static final String CONJUNCTION_OPERATOR = "and";
	
	/**
	 * Constructs a new conjunction with a given set of conjuncts.
	 * 
	 * @param arguments the conjuncts
	 */
	public Conjunction(ImmutableArray<Expression> arguments) {
		super(CONJUNCTION_OPERATOR, arguments);
	}
	
	/**
	 * Constructs a new conjunction with a given set of conjuncts.
	 * 
	 * @param arguments the conjuncts
	 */
	public Conjunction(Expression...arguments) {
		super(CONJUNCTION_OPERATOR, arguments);
	}

	@Override
	public Conjunction substitute(Substitution substitution) {
		return new Conjunction(substituteArguments(substitution));
	}
	
	@Override
	public Bindings unify(Formula other, Bindings bindings) {
		if(other instanceof Conjunction)
			return unifyArguments((NAryBooleanExpression) other, bindings);
		else
			return null;
	}

	@Override
	public boolean isTestable() {
		return true;
	}

	@Override
	public boolean isTrue(State state) {
		for(Expression argument : arguments)
			if(!argument.isTrue(state))
				return false;
		return true;
	}

	@Override
	public boolean isImposable() {
		return true;
	}
	
	@Override
	public void impose(MutableState state) {
		for(Expression argument : arguments)
			argument.impose(state);
	}

	@Override
	public ExpressionObject negate() {
		return new Disjunction(negateArguments());
	}

	@Override
	public Expression toCNF() {
		return NormalForms.toCNF(this);
	}

	@Override
	public Expression toDNF() {
		return NormalForms.toDNF(this);
	}

	@Override
	public Expression simplify() {
		if(arguments.length == 1)
			return arguments.get(0);
		else
			return new Conjunction(flatten());
	}
}

File: UnitySokoban/Assets/Scripts/Planning/Planning/Logic/Conjunction.cs
using Planning.Util;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Planning.Logic
{
    /**
     * A conjunction is a Boolean expression that is true when all of its conjuncts
     * are true.
     * 
     * @author Stephen G. Ware
     * @ported Edward Thomas Garcia
     */
    public class Conjunction : NAryBooleanExpression
    {
        /** The Boolean operator used for expressing conjunctions */
        public static readonly string CONJUNCTION_OPERATOR = "and";

        /**
	     * Constructs a new conjunction with a given set of conjuncts.
	     * 
	     * @param arguments the conjuncts
	     */
        public Conjunction(ImmutableArray<Expression> arguments) :
            base(CONJUNCTION_OPERATOR, arguments)
        {
        }

        /**
         * Constructs a new conjunction with a given set of conjuncts.
         * 
         * @param arguments the conjuncts
         */
        public Conjunction(params Expression[] arguments) :
            base(CONJUNCTION_OPERATOR, arguments)
        {
        }

        public override Expression Substitute(Substitution substitution)
        {
            return new Conjunction(SubstituteArguments(substitution));
        }

        public override Bindings Unify(Formula other, Bindings bindings)
        {
            if (other is Conjunction)
                return UnifyArguments((NAryBooleanExpression)other, bindings);
            else
                return null;
        }

        public override bool IsTestable()
        {
            return true;
        }

        public override bool IsTrue(State state)
        {
            foreach (Expression argument in arguments)
                if (!argument.IsTrue(state))
                    return false;
            return true;
        }

        public override bool IsImposable()
        {
            return true;
        }

        public override void Impose(MutableState state)
        {
            foreach (Expression argument in arguments)
                argument.Impose(state);
        }

        public override Expression Negate()
        {
            return new Disjunction(NegateArguments());
        }

        public override Expression ToCNF()
        {
            return NormalForms.toCNF(this);
        }

        public override Expression ToDNF()
        {
            return NormalForms.toDNF(this);
        }

        public override Expression Simplify()
        {
            if (arguments.length == 1)
                return arguments.get(0);
            else
                return new Conjunction(Flatten());
        }
    }
}
